File: src/main/java/edu/missouristate/mars/mips/dump/DumpFormatLoader.java
package edu.missouristate.mars.mips.dump;

import edu.missouristate.mars.util.FilenameFinder;

import java.lang.reflect.Modifier;
import java.util.ArrayList;

/****************************************************************************/
/* This class provides functionality to bring external memory dump format definitions
 * into MARS.  This is adapted from the ToolLoader class, which is in turn adapted
 * from Bret Barker's GameServer class from the book "Developing Games In Java".
 */

public class DumpFormatLoader {

    private static final String CLASS_PREFIX = "mars.mips.dump.";
    private static final String DUMP_DIRECTORY_PATH = "mars/mips/dump";
    private static final String SYSCALL_INTERFACE = "DumpFormat.class";
    private static final String CLASS_EXTENSION = "class";

    private static ArrayList<DumpFormat> formatList = null;

    /**
     * Dynamically loads dump formats into an ArrayList.  This method is adapted from
     * the loadGameControllers() method in Bret Barker's GameServer class.
     * Barker (bret@hypefiend.com) is co-author of the book "Developing Games
     * in Java".  Also see the ToolLoader and SyscallLoader classes elsewhere in MARS.
     */
    // TODO: Consider using a class indexing system with annotations instead?
    public ArrayList<DumpFormat> loadDumpFormats() {
        // The list will be populated only the first time this method is called.
        if (formatList == null) {
            formatList = new ArrayList<>();
            // grab all class files in the dump directory
            ArrayList<String> candidates = FilenameFinder.getFilenameList(this.getClass().getClassLoader(),
                    DUMP_DIRECTORY_PATH, CLASS_EXTENSION);
            for (String file : candidates) {
                try {
                    // grab the class, make sure it implements DumpFormat, instantiate, add to list
                    String formatClassName = CLASS_PREFIX + file.substring(0, file.indexOf(CLASS_EXTENSION) - 1);
                    Class<DumpFormat> clas = (Class<DumpFormat>) Class.forName(formatClassName);
                    if (DumpFormat.class.isAssignableFrom(clas) &&
                            !Modifier.isAbstract(clas.getModifiers()) &&
                            !Modifier.isInterface(clas.getModifiers())) {
                        formatList.add(clas.getDeclaredConstructor().newInstance());
                    }
                } catch (Exception e) {
                    System.out.println("Error instantiating DumpFormat from file " + file + ": " + e);
                }
            }
        }
        return formatList;
    }

    public static DumpFormat findDumpFormatGivenCommandDescriptor(ArrayList<DumpFormat> formatList, String formatCommandDescriptor) {
        DumpFormat match = null;
        for (DumpFormat dumpFormat : formatList) {
            if (dumpFormat.getCommandDescriptor().equals(formatCommandDescriptor)) {
                match = dumpFormat;
                break;
            }
        }
        return match;
    }


}

File: core/src/main/kotlin/edu/missouristate/mars/mips/dump/DumpFormatLoader.kt
/*
 * Copyright (c) 2003-2024, Pete Sanderson and Kenneth Vollmar
 * Copyright (c) 2024-present, Nicholas Hubbard
 *
 * Originally developed by Pete Sanderson (psanderson@otterbein.edu) and Kenneth Vollmar (kenvollmar@missouristate.edu)
 * Maintained by Nicholas Hubbard (nhubbard@users.noreply.github.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 *    the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.missouristate.mars.mips.dump

import edu.missouristate.mars.util.FilenameFinder
import kotlin.reflect.full.createInstance

class DumpFormatLoader {
    companion object {
        private const val CLASS_PREFIX = "edu.missouristate.mars.mips.dump."
        private const val DUMP_DIRECTORY_PATH = "edu/missouristate/mars/mipsmissouristate/mars/mips/dump"
        private const val SYSCALL_INTERFACE = "DumpFormat.class"
        private const val CLASS_EXTENSION = "class"
        private var formatList: ArrayList<DumpFormat>? = null

        /** Find a dump format from its command descriptor. */
        @JvmStatic
        fun findDumpFormatGivenCommandDescriptor(
            formatList: ArrayList<DumpFormat>,
            descriptor: String
        ): DumpFormat? = formatList.firstOrNull { it.commandDescriptor == descriptor }
    }

    /**
     * Dynamically loads dump formats into an ArrayList.
     */
    fun loadDumpFormats(): ArrayList<DumpFormat> {
        if (formatList == null) {
            formatList = arrayListOf()
            val candidates = FilenameFinder.getFilenameList(javaClass.classLoader, DUMP_DIRECTORY_PATH, CLASS_EXTENSION)
            for (file in candidates) {
                try {
                    // Grab the class, make sure it implements DumpFormat, create an instance, and add to list.
                    val formatClassName = CLASS_PREFIX + file.substringBeforeLast(CLASS_EXTENSION).dropLast(1)
                    val kClass = Class.forName(formatClassName).kotlin
                    if (DumpFormat::class.java.isAssignableFrom(kClass.java) && !kClass.isAbstract && !kClass.java.isInterface) {
                        // Create instance and add to list
                        val instance = kClass.createInstance() as DumpFormat
                        formatList!!.add(instance)
                    }
                } catch (e: Exception) {
                    println("Error instantiating DumpFormat from file $file: ${e.message}")
                }
            }
        }
        return formatList!!
    }
}